iree/hal/local/loaders/wasm_module_loader.c

```diff
@@ -0,0 +1,552 @@
+  // sandboxed runtime unless explicitly allowed through module imports.
+
+  // TODO(#4022): Use a smarter allocator / allocation scheme.
+  //   This is just MVP placeholder code.
+
+  // Deep clone dispatch_state and workgroup_id using WASM-accessible memory.
+  // All pointers in these structs should be WASM virtual memory addresses, not
+  // the typical host pointers or memory addresses.
+
+  // Clone dispatch_state.
+  void* native_buffer_dispatch_state = NULL;
+  uint32_t wasm_buffer_dispatch_state = wasm_runtime_module_malloc(
+      executable->wasm_module_inst,
+      sizeof(iree_hal_executable_dispatch_state_v0_t),
+      &native_buffer_dispatch_state);
+  if (IREE_UNLIKELY(wasm_buffer_dispatch_state == 0)) {
+    return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED, "malloc failed");
+  }
+  iree_hal_executable_dispatch_state_v0_t* wasm_dispatch_state =
+      (iree_hal_executable_dispatch_state_v0_t*)native_buffer_dispatch_state;
+  // Workground count/size.
+  wasm_dispatch_state->workgroup_count.x = dispatch_state->workgroup_count.x;
+  wasm_dispatch_state->workgroup_count.y = dispatch_state->workgroup_count.y;
+  wasm_dispatch_state->workgroup_count.z = dispatch_state->workgroup_count.z;
+  wasm_dispatch_state->workgroup_size.x = dispatch_state->workgroup_size.x;
+  wasm_dispatch_state->workgroup_size.y = dispatch_state->workgroup_size.y;
+  wasm_dispatch_state->workgroup_size.z = dispatch_state->workgroup_size.z;
+  // Push constants.
+  wasm_dispatch_state->push_constant_count =
+      dispatch_state->push_constant_count;
+  void* native_buffer_push_constants = NULL;
+  uint32_t wasm_buffer_push_constants = wasm_runtime_module_malloc(
+      executable->wasm_module_inst,
+      dispatch_state->push_constant_count * sizeof(int32_t*),
+      &native_buffer_push_constants);
+  if (IREE_UNLIKELY(wasm_buffer_push_constants == 0)) {
+    return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED, "malloc failed");
+  }
+  int32_t* wasm_push_constants = (int32_t*)native_buffer_push_constants;
+  for (int i = 0; i < dispatch_state->push_constant_count; ++i) {
+    wasm_push_constants[i] = dispatch_state->push_constants[i];
+  }
+  // TODO(scotttodd): fix:
+  //   '=': 'const uint32_t *' differs in levels of indirection from 'uint32_t'
+  wasm_dispatch_state->push_constants = wasm_buffer_push_constants;
+  // Bindings.
+  wasm_dispatch_state->binding_count = dispatch_state->binding_count;
+  //   - binding_ptrs
+  void* native_buffer_binding_ptrs = NULL;
+  uint32_t wasm_buffer_binding_ptrs = wasm_runtime_module_malloc(
+      executable->wasm_module_inst,
+      dispatch_state->binding_count * sizeof(int32_t*),
+      &native_buffer_binding_ptrs);
+  if (IREE_UNLIKELY(wasm_buffer_binding_ptrs == 0)) {
+    return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED, "malloc failed");
+  }
+  int32_t* wasm_binding_ptrs = (int32_t*)native_buffer_binding_ptrs;
+  // HACK: remember the last native buffer (usually the single output buffer)
```

Comment:
This is the scary one :P
Every single fiber of my being is telling me to say this should be fixed now; the stack stuff can be fixed later (along with threading), but this is a biggie. It's worth implementing the HAL allocator to malloc/free from the wasm heap so all of this can be avoided. MVPs should at least approximate the final state.

If there's some bigger blockers preventing that I could be convinced, but I suspect there aren't (given that cuda/vulkan/etc work with their weird allocators). Worth at least copy/pasting the file and giving it a shot now.

Reply:
The other big thing this will force is how to share memory across the wasm modules - which may require some reworking of the ownership here as the loader (or HAL allocator) may need to own some wamr things, vs each executable owning the entire world. We don't want to get into the state where you can only ever have a single executable loaded per HAL device, and currently that would be the case with a naive extension (the HAL allocator, if created independent of the loader, won't be able to share memory across executables).

Reply:
Ok, looking into this. First step is figuring out how to set a different `device_allocator` on `task_device`. That isn't configurable right now, and is set here: https://github.com/google/iree/blob/51b6453e5772ecaa92e65ff6a85ce7f32a9c7869/iree/hal/local/task_device.c#L141-L144

Reply:
<sub>(maybe move discussion to Discord?)</sub>

How about adding a (optional?) factory function for creating a `device_allocator` to `iree_hal_task_device_params_t`, which is passed to `iree_hal_task_driver_create`? I can give that a try, at least to get started on writing a WASM allocator. 
https://github.com/google/iree/blob/51b6453e5772ecaa92e65ff6a85ce7f32a9c7869/iree/hal/local/task_device.h#L27-L39

Reply:
Some discussion moved to Discord [here](https://discord.com/channels/689900678990135345/782059441641881630/821157599679479808).

Still figuring out an implementation path.